### PR TITLE
graph publication column save type should be json instead of string

### LINKED ISSUE #8172: Cached graph is saved to JSONField as a string, should be dictionary/object
We save a serialized version of the graph models to `graph_publications.serialized_graph`, which is a JSONField.  Currently, this is stored as a string, which causes some issues with how we handle the serialized graph (see the add resource page).  It also would be more useful/usable in the database if stored as an object.

When we assign this column, it should be using a dictionary so that we store a proper JSON object.

### PR DESCRIPTION
**Pull request by aarongundel**
### Types of changes

-   [x] Bugfix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Description of Change

Fix for issue where a string based index was expected, but a string object was received.  This happened because the graph was saved to the db as a string instead of a JSON structure

### Issues Solved

Fixes #8172 

### Checklist

-   [x] Unit tests pass locally with my changes
-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [ ] I have added necessary documentation (if appropriate)

#### Ticket Background
*   Sponsored by: 
*   Found by: @ 
*   Tested by: @ 
*   Designed by: @

### CODE REVIEW

**Comment by aarongundel** [comment]
Note that in order to test you should roll back to 8085 and then roll forward - this will re-save the cached items as json structured items instead of strings.

**Comment by aarongundel** [comment]
@robgaston requested more eyes since he's occupied with training until tomorrow

File: arches/app/models/graph.py [hunk 1]
```diff
@@ -1307,7 +1307,7 @@ def serialize(self, fields=None, exclude=None, force_recalculation=False):
         exclude = [] if exclude is None else exclude
 
         if self.publication and not force_recalculation:
-            deserialized_graph = JSONDeserializer().deserialize(self.publication.serialized_graph)
+            deserialized_graph = self.publication.serialized_graph
```
Comment (chrabyrd) [inline arches/app/models/graph.py:1310]: nit: can `deserialized_graph` be changed to something else?  `deserialized_graph = ...serialized_graph` is confusing

File: arches/app/models/graph.py [hunk 2]
```diff
@@ -1586,15 +1589,18 @@ def publish(self, notes=None):
...
+                notes=notes,
+            )
+            publication.save()
 
-        self.publication = publication
-        self.save()
+            self.publication = publication
+            self.save()
+        except Exception:
+            raise UnpublishedModelError("Could not save graph publication ")
```
Comment (chrabyrd) [inline arches/app/models/graph.py:1603]: can this error be more descriptive, can we leverage the exception?

File: arches/app/models/graph.py [hunk 3]
```diff
@@ -1586,15 +1589,18 @@ def publish(self, notes=None):
         Adds a row to the GraphPublication table
         Assigns GraphPublication id to Graph
         """
-        publication = models.GraphPublication.objects.create(
-            graph=self,
-            serialized_graph=JSONSerializer().serialize(self, force_recalculation=True),
-            notes=notes,
-        )
-        publication.save()
+        try:
```
Comment (chrabyrd) [inline arches/app/models/graph.py:1592]: when will this fail? is it possible / worth it to replace the try/except with a check of some kind?

File: arches/app/models/migrations/7784_add_serialized_graph_to_graph_publications.py [hunk 4]
```diff
@@ -14,7 +14,9 @@ class Migration(migrations.Migration):
     def forwards_add_serialized_graph_column_data(apps, schema_editor):
         for graph in Graph.objects.all():
             if graph.publication:
-                graph.publication.serialized_graph = JSONSerializer().serialize(graph, force_recalculation=True)
+                graph.publication.serialized_graph = JSONDeserializer().deserialize(
```
Comment (chrabyrd) [inline arches/app/models/migrations/7784_add_serialized_graph_to_graph_publications.py:17]: is there any point to saving a deserialized serialized graph? Can the original graph just be used?

**Review by chrabyrd — COMMENTED** [review]
overall looks good just some minor questions 👍

File: arches/app/models/graph.py [hunk 5]
```diff
@@ -1377,14 +1377,17 @@ def find_unpermitted_edits(obj_a, obj_b, ignore_list, obj_type):
...
+                    try:
+                        unpermitted_node_edits = find_unpermitted_edits(
+                            db_node,
+                            self.nodes[db_node.nodeid],
+                            ["name", "issearchable", "ontologyclass", "description", "isrequired", "fieldname", "exportable"],
+                            "node",
+                        )
+                        if unpermitted_node_edits is not None:
+                            unpermitted_edits.append(unpermitted_node_edits)
+                    except KeyError:
```
Comment (chrabyrd) [inline arches/app/models/graph.py:1389]: what causes this to error?

File: arches/app/models/graph.py (hunk 1 above)
Comment (aarongundel) [reply arches/app/models/graph.py:1310]: yep it is.

File: arches/app/models/migrations/7784_add_serialized_graph_to_graph_publications.py (hunk 4 above)
Comment (aarongundel) [reply arches/app/models/migrations/7784_add_serialized_graph_to_graph_publications.py:17]: No; the original graph isn't compatible with the JsonField type and blows up.

Comment (chrabyrd) [reply arches/app/models/migrations/7784_add_serialized_graph_to_graph_publications.py:17]: cool just wanted to check 👍

File: arches/app/models/graph.py (hunk 5 above)
Comment (aarongundel) [reply arches/app/models/graph.py:1389]: This was actually the result of a corrupt graph I had - where there was no db_node.nodeid in self.nodes.  I'll remove this; as it should never happen.

File: arches/app/models/graph.py (hunk 2 above)
Comment (aarongundel) [reply arches/app/models/graph.py]: not sure how much more descriptive it can be; happy to get your thoughts on this.

Comment (chrabyrd) [reply arches/app/models/graph.py]: oh sure, instead of that sentence can pass the declared Exception? `except Exception as e:`

**Review by robgaston — CHANGES_REQUESTED** [review]
this looks good to me, I think we just need to resolve the merge conflicts in graph.py before this will be ready to go

File: arches/app/models/graph.py [hunk 6]
```diff
@@ -1586,15 +1586,22 @@ def publish(self, notes=None):
...
+        with transaction.atomic():
+            try:
+                publication = models.GraphPublication.objects.create(
+                    graph=self,
+                    serialized_graph=JSONDeserializer().deserialize(JSONSerializer().serialize(self, force_recalculation=True)),
+                    notes=notes,
+                )
+                publication.save()
+            except Exception:
+                raise UnpublishedModelError("Could not save graph publication")
```
Comment (chrabyrd) [inline arches/app/models/graph.py:1598]: Instead of raising the string, can we pass the exception here?

File: arches/app/models/graph.py [hunk 7]
```diff
@@ -1586,15 +1586,22 @@ def publish(self, notes=None):
...
 
-        self.publication = publication
-        self.save()
+            try:
+                self.publication = publication
+                self.save()
+            except Exception:
+                raise UnpublishedModelError("Could not save graph with new publication")
```
Comment (chrabyrd) [inline arches/app/models/graph.py:1604]: Instead of raising the string, can we pass the exception here?

**Review by chrabyrd — COMMENTED** [review]
Looks great just a minor nit 👍

File: arches/app/models/graph.py (hunk 6 above)
Comment (aarongundel) [reply arches/app/models/graph.py]: good feedback, done and done.

**Review by robgaston — APPROVED** [review]
looks good

**Pull request merged**